tools/dbus-send.1
.\" 
.\" dbus-send manual page.
.\" Copyright (C) 2003 Red Hat, Inc.
.\"
.TH dbus-send 1
.SH NAME
dbus-send \- Send a message to a message bus
.SH SYNOPSIS
.PP
.B dbus-send
[\-\-system | \-\-session] [\-\-dest=SERVICE] [\-\-print-reply] <message name> [contents ...]

.SH DESCRIPTION

The \fIdbus-send\fP command is used to send a message to a D-BUS message
bus. See http://www.freedesktop.org/software/dbus/ for more 
information about the big picture.

.PP
There are two well-known message buses: the systemwide message bus 
(installed on many systems as the "messagebus" service) and the 
per-user-login-session message bus (started each time a user logs in).
The \-\-system and \-\-session options direct \fIdbus-send\fP to send
messages to the system or session buses respectively.  If neither is
specified, \fIdbus-send\fP sends to the session bus.

.PP 
Nearly all uses of \fIdbus-send\fP must provide the \-\-dest argument
which is the name of a service on the bus to send the message to. If
\-\-dest is omitted, a default service name of
"org.freedesktop.DBus.Broadcast" is used.  

.PP
The name of the message to send must always be specified. Following
arguments, if any, are the message contents (message arguments).
These are given as a type name, a colon, and then the value of the
argument. The possible type names are: string, int32, uint32, double,
byte, boolean.  (D-BUS supports more types than these, but
\fIdbus-send\fP currently does not.)

.PP
Here is an example invocation:
.nf

  dbus-send \-\-dest='org.freedesktop.ExampleService'     \\
            org.freedesktop.ExampleMessage              \\
            int32:47 string:'hello world' double:65.32

.fi

.SH OPTIONS
The following options are supported:
.TP
.I "--dest=SERVICE"
Specify the service to receive the message.
.TP
.I "--print-reply"
Block for a reply to the message sent, and print any reply received.
.TP
.I "--system"
Send to the system message bus.
.TP
.I "--session"
Send to the session message bus.  (This is the default.)

.SH AUTHOR
dbus-send was written by Philip Blundell.

.SH BUGS
Please send bug reports to the D-BUS mailing list or bug tracker,
see http://www.freedesktop.org/software/dbus/


NEWS
D-BUS 0.12
===

- fix "service messagebus status" on Red Hat
- fix demarshaling of DBUS_TYPE_NAMED
- fix "eval `dbus-launch --exit-with-session`" to exit properly
- fix build without --prefix
- useless fooling with Mono bindings
- useless fooling with gcj bindings
- fix srcdir != builddir
- fix various compiler warnings and other issues
- add get/set data to DBusMessage
- fix headers for C++
- OS X build fixes
- abstract domain sockets support (Linux only)
- add dbus-cleanup-sockets utility for people
  not using linux
- be consistent about defaulting to --session/--system 
  with command line tools
- merge in policies from included config files
- fix build on non-x86
- docs updates
- lots of other bugfixes

D-BUS 0.11
===

- add --enable-docs to turn off/on the docbook stuff
  (doesn't cover doxygen stuff yet)
- make people define DBUS_API_SUBJECT_TO_CHANGE
  and read warning in README so they don't 
  expect the API to be frozen already
- rename .pc files to "dbus-1.pc" instead of 
  "dbus-1.0.pc" etc. - this will require changing 
  pkg-config invocations
- some docs cleanups
- add man pages for all executables
- allow send to/from bus driver in the default system.conf
- fix user lookup bug
- implement dbus-launch to launch the session message bus
- fix some thread deadlocks
- some performance profiling/optimization
- add dbus_bus_activate_service() function
- fix some minor bugs here and there
- install Red Hat initscript in the right place

D-BUS 0.10
===

- reversed order of args to dbus_message_new()
- renamed dbus_message_name_is() and some other
  functions
- change DBusWatch to have dbus_watch_handle() 
  similar to dbus_timeout_handle(), drop 
  connection/server-specific handle routines
- change message serials to be unsigned
- implemented <allow>/<deny>/<limit> features for 
  config file; system bus now has a deny-all policy 
  by default.
- system.conf has <includedir>system.d</includedir>
  so packages can install additions to the default 
  policy to <allow> the messages they need. e.g.
  CUPS might install a cups.conf - see 
  test/data/valid-config-files/system.d/test.conf 
  for an example.
- add timeouts for authentication, activation
- add glib-style "checks" on public API, enable 
  those by default, disable assertions by default
- add GMainContext argument to GLib setup functions, 
  can be NULL for default context. Needed for threads.
- add 64-bit integer type
- validate type of standard message header fields
- consider messages in the org.freedesktop.Local
  namespace to be invalid (to avoid fake disconnect 
  messages for example)
- fix assorted memory leaks and other bugs in 
  the SHA-1 auth mechanism
- cache user database information (groups user is 
  in, etc.) helps a lot with NIS
- always store uid_t, pid_t, gid_t in "ulong" 
  rather than "int"
- implement config file settings for which 
  users can connect
- SHA-1 unit test
- dbus-send, dbus-monitor command line utilities
- fixed lots of misc crashes and other bugs

D-BUS 0.9
===

- implemented a test case for service activation,
  and fixed many bugs exposed by that
- implemented recursive argument marshaling/demarshaling 
  for messages, allowing multidimensional arrays
- fixed up integration of message dispatch with 
  main loop by adding a callback on change of 
  dispatch status
- add a pidfile feature to daemon
- some build fixes
- clean up unix domain sockets on exit
- add --print-address and the ability 
  to create a random server address 
  in a temporary directory

D-BUS 0.8
===

- fix dumb bug in 0.7

D-BUS 0.7
===

- implement configuration file used to control bus characteristics
- implement daemon mode, changing user ID, and other system
  bus features
- add init scripts for systemwide bus
- add "make check-coverage" target to check test coverage
- more test suite additions
- many, many bugfixes
- many API changes/fixes

D-BUS 0.6
===
- Vastly improved bus daemon test suite
- Lots of misc. bugfixes and memory leak fixes
- Support for marshalling key/value data
- Activation improvements.

D-BUS 0.5
===

- Specification updates
- port to OS X and other BSD variants
- port to Solaris
- Partial work on cookie-based authentication
- Thread safety fixes
- Lots of misc. bugfixes
- Support for more array types
- Add data slots to DBusServer
- DBusString security audit fixes
- Fix for systems (or valgrind) with unaligned malloc blocks

D-BUS 0.4
===

- Preliminary activation support.
- Better authentication test suite
- Bus test program
- Specification updates
- Thread safety
- Bug fixes

D-BUS 0.3
===

 - Preliminary limitations 
 - Message sending works
 - Bus client
 - Array marshalling/demarshalling
 - Services
 - Better OOM handling in the bus
 - In-proc debug transport
 - Transport/server address support

D-BUS 0.2
===

 - Include test code in the tarball.

D-BUS 0.1
===

 - Initial release.


tools/dbus-send.c
/* -*- mode: C; c-file-style: "gnu" -*- */
/* dbus-send.c  Utility program to send messages from the command line
 *
 * Copyright (C) 2003 Philip Blundell <philb@gnu.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <dbus/dbus.h>

#include "dbus-print-message.h"

static void
usage (char *name, int ecode)
{
  fprintf (stderr, "Usage: %s [--help] [--system | --session] [--dest=SERVICE] [--type=TYPE] [--print-reply] [--reply-timeout=MSEC] <destination object path> <message name> [contents ...]\n", name);
  exit (ecode);
}

int
main (int argc, char *argv[])
{
  DBusConnection *connection;
  DBusError error;
  DBusMessage *message;
  int print_reply;
  int reply_timeout;
  DBusMessageIter iter;
  int i;
  DBusBusType type = DBUS_BUS_SESSION;
  const char *dest = NULL;
  const char *name = NULL;
  const char *path = NULL;
  int message_type = DBUS_MESSAGE_TYPE_SIGNAL;
  const char *type_str = NULL;
  
  if (argc < 3)
    usage (argv[0], 1);

  print_reply = FALSE;
  reply_timeout = -1;
  
  for (i = 1; i < argc && name == NULL; i++)
    {
      char *arg = argv[i];

      if (strcmp (arg, "--system") == 0)
	type = DBUS_BUS_SYSTEM;
      else if (strcmp (arg, "--session") == 0)
	type = DBUS_BUS_SESSION;
      else if (strcmp (arg, "--print-reply") == 0)
	{
	  print_reply = TRUE;
	  message_type = DBUS_MESSAGE_TYPE_METHOD_CALL;
	}
      else if (strstr (arg, "--reply-timeout=") == arg)
	{
	  reply_timeout = strtol (strchr (arg, '=') + 1,
				  NULL, 10);
	}
      else if (strstr (arg, "--dest=") == arg)
	dest = strchr (arg, '=') + 1;
      else if (strstr (arg, "--type=") == arg)
	type_str = strchr (arg, '=') + 1;
      else if (!strcmp(arg, "--help"))
	usage (argv[0], 0);
      else if (arg[0] == '-')
	usage (argv[0], 1);
      else if (path == NULL)
        path = arg;
      else if (name == NULL)
        name = arg;
      else
        usage (argv[0], 1);
    }

  if (name == NULL)
    usage (argv[0], 1);

  if (type_str != NULL)
    {
      message_type = dbus_message_type_from_string (type_str);
      if (!(message_type == DBUS_MESSAGE_TYPE_METHOD_CALL ||
            message_type == DBUS_MESSAGE_TYPE_SIGNAL))
        {
          fprintf (stderr, "Message type \"%s\" is not supported\n",
                   type_str);
          exit (1);
        }
    }
  
  dbus_error_init (&error);
  connection = dbus_bus_get (type, &error);
  if (connection == NULL)
    {
      fprintf (stderr, "Failed to open connection to %s message bus: %s\n",
	       (type == DBUS_BUS_SYSTEM) ? "system" : "session",
               error.message);
      dbus_error_free (&error);
      exit (1);
    }

  if (message_type == DBUS_MESSAGE_TYPE_METHOD_CALL)
    {
      char *last_dot;

      last_dot = strrchr (name, '.');
      if (last_dot == NULL)
        {
          fprintf (stderr, "Must use org.mydomain.Interface.Method notation, no dot in \"%s\"\n",
                   name);
          exit (1);
        }
      *last_dot = '\0';
      
      message = dbus_message_new_method_call (NULL,
                                              path,
                                              name,
                                              last_dot + 1);
    }
  else if (message_type == DBUS_MESSAGE_TYPE_SIGNAL)
    {
      char *last_dot;

      last_dot = strrchr (name, '.');
      if (last_dot == NULL)
        {
          fprintf (stderr, "Must use org.mydomain.Interface.Signal notation, no dot in \"%s\"\n",
                   name);
          exit (1);
        }
      *last_dot = '\0';
      
      message = dbus_message_new_signal (path, name, last_dot + 1);
    }
  else
    {
      fprintf (stderr, "Internal error, unknown message type\n");
      exit (1);
    }

  if (message == NULL)
    {
      fprintf (stderr, "Couldn't allocate D-BUS message\n");
      exit (1);
    }

  if (dest && !dbus_message_set_destination (message, dest))
    {
      fprintf (stderr, "Not enough memory\n");
      exit (1);
    }
  
  dbus_message_append_iter_init (message, &iter);

  while (i < argc)
    {
      char *arg;
      char *c;
      int type;
      dbus_uint32_t uint32;
      dbus_int32_t int32;
      double d;
      unsigned char byte;

      type = DBUS_TYPE_INVALID;
      arg = argv[i++];
      c = strchr (arg, ':');

      if (c == NULL)
	{
	  fprintf (stderr, "%s: Data item \"%s\" is badly formed\n", argv[0], arg);
	  exit (1);
	}

      *(c++) = 0;

      if (arg[0] == 0 || !strcmp (arg, "string"))
	type = DBUS_TYPE_STRING;
      else if (!strcmp (arg, "int32"))
	type = DBUS_TYPE_INT32;
      else if (!strcmp (arg, "uint32"))
	type = DBUS_TYPE_UINT32;
      else if (!strcmp (arg, "double"))
	type = DBUS_TYPE_DOUBLE;
      else if (!strcmp (arg, "byte"))
	type = DBUS_TYPE_BYTE;
      else if (!strcmp (arg, "boolean"))
	type = DBUS_TYPE_BOOLEAN;
      else
	{
	  fprintf (stderr, "%s: Unknown type \"%s\"\n", argv[0], arg);
	  exit (1);
	}

      /* FIXME - we are ignoring OOM returns on all these functions */
      switch (type)
	{
	case DBUS_TYPE_BYTE:
	  byte = strtoul (c, NULL, 0);
	  dbus_message_iter_append_basic (&iter, DBUS_TYPE_BYTE, &byte);
	  break;

	case DBUS_TYPE_DOUBLE:
	  d = strtod (c, NULL);
	  dbus_message_iter_append_basic (&iter, DBUS_TYPE_DOUBLE, &d);
	  break;

	case DBUS_TYPE_INT32:
	  int32 = strtol (c, NULL, 0);
	  dbus_message_iter_append_basic (&iter, DBUS_TYPE_INT32, &int32);
	  break;

	case DBUS_TYPE_UINT32:
	  uint32 = strtoul (c, NULL, 0);
	  dbus_message_iter_append_basic (&iter, DBUS_TYPE_UINT32, &uint32);
	  break;

	case DBUS_TYPE_STRING:
	  dbus_message_iter_append_basic (&iter, DBUS_TYPE_STRING, &c);
	  break;

	case DBUS_TYPE_BOOLEAN:
          if (strcmp(c, "true") == 0)
            {
              byte = TRUE;
              dbus_message_iter_append_basic (&iter, DBUS_TYPE_BOOLEAN, &c);
            }
	  else if (strcmp(c, "false") == 0)
            {
              byte = FALSE;
              dbus_message_iter_append_basic (&iter, DBUS_TYPE_BOOLEAN, &c);
            }
	  else
	    {
	      fprintf (stderr, "%s: Expected \"true\" or \"false\" instead of \"%s\"\n", argv[0], c);
	      exit (1);
	    }
	  break;

	default:
	  fprintf (stderr, "%s: Unsupported data type\n", argv[0]);
	  exit (1);
	}
    }

  if (print_reply)
    {
      DBusMessage *reply;

      dbus_error_init (&error);
      reply = dbus_connection_send_with_reply_and_block (connection,
                                                         message, reply_timeout,
                                                         &error);
      if (dbus_error_is_set (&error))
        {
          fprintf (stderr, "Error: %s\n",
                   error.message);
          exit (1);
        }

      if (reply)
        {
          print_message (reply);
          dbus_message_unref (reply);
        }
    }
  else
    {
      dbus_connection_send (connection, message, NULL);
      dbus_connection_flush (connection);
    }

  dbus_message_unref (message);

  dbus_connection_disconnect (connection);

  exit (0);
}
